### PR TITLE
feat: set FlightDescriptor on FlightDataEncoderBuilder

### LINKED ISSUE #3855: Enable setting FlightDescriptor on FlightDataEncoderBuilder
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
When using FlightDataEncoder to send arrow batches to a flight do_put endpoint, FlightDescriptor cannot easily be set on first message. https://arrow.apache.org/docs/format/Flight.html#uploading-data

**Describe the solution you'd like**
On FlightDataEncoderBuilder a `with_flight_descriptor` method would be nice. The FlightDescriptor would then be set on the first FlightData retrieved from the encoder.

**Describe alternatives you've considered**
Here is a workaround, I may be missing something simpler
```rust
let mut encoder = FlightDataEncoderBuilder::new()
    .with_schema(schema.clone())
    .build(stream);

let mut first_flight_data = encoder
    .next()
    .await
    .expect("no flight data")
    .expect("flight data was error");
first_flight_data.flight_descriptor = Some(FlightDescriptor {
    r#type: DescriptorType::Path.into(),
    path: vec!["table_name".to_string()],
    cmd: Bytes::default(),
});
let put_stream = stream::once(async { Ok(first_flight_data) }).chain(encoder);
```

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #3855

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-flight/src/encode.rs [hunk 1]
```diff
@@ -153,13 +165,15 @@ impl FlightDataEncoderBuilder {
...
             app_metadata,
+            descriptor,
         )
     }
 }
 
 /// Stream that encodes a stream of record batches to flight data.
 ///
 /// See [`FlightDataEncoderBuilder`] for details and example.
+#[warn(dead_code)]
```
Comment (tustvold) [inline arrow-flight/src/encode.rs:176]: Why this change?

File: arrow-flight/src/encode.rs [hunk 2]
```diff
@@ -210,7 +228,13 @@ impl FlightDataEncoder {
 
     /// Place the `FlightData` in the queue to send
     fn queue_messages(&mut self, datas: impl IntoIterator<Item = FlightData>) {
-        for data in datas {
+        let mut is_first = true;
+        for mut data in datas {
+            // The first message is the schema message need to set the descriptor
+            if is_first {
+                data.flight_descriptor = self.descriptor.clone();
```
Comment (tustvold) [inline arrow-flight/src/encode.rs:235]: I think the ticket calls for sending a separate descriptor message, not overriding the first message

**Comment by Weijun-H** [comment]
Error happened, but not by my pr
```
Run cargo fmt --all -- --check
Diff in /__w/arrow-rs/arrow-rs/arrow-string/src/like.rs at line 1244:
             "ﬀkoß",
             "😃sadlksFFkoSSsh😃klF", // Original was case insensitive "😃sadlksffkosSsh😃klF"
             "😱slgFFkoSSsh😃klF",    // Original was case insensitive "😱slgffkosSsh😃klF"
-            "FFkoSS",                    // "FFKoSS"
+            "FFkoSS",                // "FFKoSS"
         ],
         "FFkoSS",
         contains_utf[8](https://github.com/apache/arrow-rs/actions/runs/4758784682/jobs/8457275692?pr=4101#step:6:9)_scalar,
```

File: arrow-flight/src/encode.rs [hunk 3]
```diff
@@ -134,6 +137,14 @@ impl FlightDataEncoderBuilder {
         self
     }
 
+    pub fn with_flight_descriptor(
```
Comment (alamb) [inline arrow-flight/src/encode.rs:140]: Can you please add doc strings to explain how this function works, to keep in line with the rest of the functions on this struct?

Specifically that it is sent on the first FlightData as described in https://arrow.apache.org/docs/format/Flight.html#uploading-data

File: arrow-flight/src/encode.rs [hunk 4]
```diff
@@ -233,6 +256,24 @@ impl FlightDataEncoder {
...
+        let descriptor = descriptor.clone();
+        let descriptor_flight_data = FlightData::new(
+            Some(descriptor.clone()),
+            IpcMessage(Bytes::new()),
+            vec![],
+            vec![],
+        );
+        self.queue_message(descriptor_flight_data);
+        // remember descriptor
+        self.descriptor = Some(descriptor.clone());
```
Comment (alamb) [inline arrow-flight/src/encode.rs:273]: Why is the descriptor remembered?

File: arrow-flight/src/encode.rs [hunk 5]
```diff
@@ -176,6 +189,8 @@ pub struct FlightDataEncoder {
     queue: VecDeque<FlightData>,
     /// Is this stream done (inner is empty or errored)
     done: bool,
+    /// descriptor, set after the first batch
```
Comment (alamb) [inline arrow-flight/src/encode.rs:192]: What does "set after the first batch" mean? 

I guess I expect this to be something more like "cleared after the first FlightData message"

File: arrow-flight/src/encode.rs [hunk 6]
```diff
@@ -194,12 +210,19 @@ impl FlightDataEncoder {
...
+            descriptor: None,
         };
 
         // If schema is known up front, enqueue it immediately
         if let Some(schema) = schema {
             encoder.encode_schema(&schema);
         }
+
+        // If descriptor is known up front, enqueue it immediately
+        if let Some(descriptor) = descriptor {
```
Comment (alamb) [inline arrow-flight/src/encode.rs:222]: As I understand the intent, the idea is that the descriptor is set on the first message that is sent this code will potentially send a separate message after the schema message, if the schema is known up front.

Perhaps you could check  `self.descriptor` in `enqueue_messages` and attach the descriptor there

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H 

I think this PR needs: 
1. A test of this feature (as I don't think it is working as expected, but I may be mistaken. A test will show us one way or the other). Perhaps in 
https://github.com/apache/arrow-rs/blob/master/arrow-flight/tests/encode_decode.rs
2. Some documentation describing when the flight descriptor is sent

File: arrow-flight/src/encode.rs (hunk 3 above)
Comment (Weijun-H) [reply arrow-flight/src/encode.rs:141]: I am not sure if the data is empty, should we also need to send a seperate empty FligtData with the descriptor.

Comment (alamb) [reply arrow-flight/src/encode.rs:141]: If the encoder would not have sent any `FlightData` if no `FlightDescriptor` is provided, then I don't think a `FlightData` should be sent if a `FlightDescriptor` is sent)

File: arrow-flight/src/encode.rs [hunk 7]
```diff
@@ -194,17 +211,28 @@ impl FlightDataEncoder {
...
             done: false,
+            descriptor: None,
         };
 
         // If schema is known up front, enqueue it immediately
         if let Some(schema) = schema {
             encoder.encode_schema(&schema);
         }
+
+        encoder.descriptor = descriptor;
```
Comment (alamb) [inline arrow-flight/src/encode.rs:222]: I think this should be set before calling `encoder.encode_schema()` so it is included on the first Schema message, if provided

File: arrow-flight/src/encode.rs [hunk 8]
```diff
@@ -194,17 +211,28 @@ impl FlightDataEncoder {
...
     fn queue_message(&mut self, data: FlightData) {
+        if let Some(descriptor) = &self.descriptor {
+            let mut data = data;
+            data.flight_descriptor = Some(descriptor.clone());
+            self.descriptor = None;
+            self.queue.push_back(data);
+            return;
+        }
         self.queue.push_back(data);
     }
```
Comment (alamb) [inline arrow-flight/src/encode.rs:237]: FWIW I think you can write this using `take()` a bit more concisely like:

```suggestion
    fn queue_message(&mut self, mut data: FlightData) {
        if let Some(descriptor) = self.descriptor.take() {
            data.flight_descriptor = Some(descriptor.clone());
        }
        self.queue.push_back(data);
    }
```

File: arrow-flight/tests/encode_decode.rs [hunk 9]
```diff
@@ -136,6 +138,31 @@ async fn test_zero_batches_schema_specified() {
...
+    });
+
+    let encoder = FlightDataEncoderBuilder::default()
+        .with_schema(schema.clone())
+        .with_flight_descriptor(descriptor.clone());
+
+    let mut encoder = encoder.build(stream);
+
+    // First batch should be the schema
+    encoder.next().await.unwrap().unwrap();
```
Comment (alamb) [inline arrow-flight/tests/encode_decode.rs:159]: My understanding of https://arrow.apache.org/docs/format/Flight.html#uploading-data was that the first `FlightData` should have the `descriptor`, not the second 🤔

**Review by alamb — COMMENTED** [review]
THanks @Weijun-H  -- this is looking close

File: arrow-flight/tests/encode_decode.rs (hunk 9 above)
Comment (Weijun-H) [reply arrow-flight/tests/encode_decode.rs]: Yes, you are right. I completely misunderstood it.

**Review by alamb — APPROVED** [review]
Look good -- thank you @Weijun-H
